[ruiyigan] iP

ruiyigan opened the pull request:
# Duke Super Gan 🚀
_A very simple application_
## Description
Duke Super Gan is a task manager, best used without a mouse. 

> **_Why click, when you can type?_**

## Features ✅
- Addition of 3 different types of tasks: `todo` |  `event` | `deadline`
- Find task simply by using the `find` command
- Mark or unmark your tasks by typing `mark` or `unmark`
- Delete useless tasks simply by hitting `remove`
- Persistent storage. Your data will be here when you're back.

## Upcoming features ⚡️
- [ ] Reminders
- [ ] Statistics

## Usage 🛠️
1. Download it [here](https://github.com/ruiyigan/ip/releases/download/A-Jar/duke.jar)
2. Copy the jar file into an empty folder
3. Open a command window in that folder
4. Run the command java -jar {filename}.jar e.g., java -jar Duke.jar (i.e., run the command in the same folder as the jar file)

## Developer Guide
If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Duke {
    public static void main(String[] args) {
        new Duke(<location of local storage file in string>).run();
    }
}
```

Commit:
Add Gradle support

ruiyigan committed:
added logic for level 1 Duke.java

ruiyigan committed:
added comments on Duke.java for better documentation

ruiyigan committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

ruiyigan committed:
Added Task.class

ruiyigan committed:
Add the ability to mark tasks as done and the ability to change the status back to not done.

ruiyigan committed:
Prevent index out of range issue

ruiyigan committed:
Implemented inheristance

ruiyigan committed:
Add support for tracking three types of tasks

ruiyigan committed:
Used static for class counter instead of declaring counter

ruiyigan committed:
Use the input/output redirection technique to semi-automate the testing of Duke.

ruiyigan committed:
Create DukeException and added first exception from input other than specific 3 types of inputs

ruiyigan committed:
Handled Errors

ruiyigan committed:
Switched to ArrayList

ruiyigan committed:
Add support for deleting tasks from the list.

ruiyigan committed:
Used java split instead of contains to simplify logic

ruiyigan committed:
Abstracted methods into TaskList array

ruiyigan committed:
Add saving tasks in the hard disk

ruiyigan committed:
Add function to load data from hard disk when Duke starts up

ruiyigan committed:
Taught Duke how to understand dates and times

ruiyigan committed:
Resolve conflict from branch-Level-7

ruiyigan committed:
Fix bugs in Storage.java

ruiyigan committed:
Fix spacing issue

ruiyigan committed:
Merge without fast forwardingMerge branch 'branch-Level-7'

ruiyigan committed:
Fix spacing issue with saving data

ruiyigan committed:
Merge branch 'branch-Level-7'

ruiyigan committed:
Update .gitignore to ignore data folder

ruiyigan committed:
Abstracted different functionalities to different classes

ruiyigan committed:
Add Command.java to abstract commands execution

ruiyigan committed:
Divide classes into packages

ruiyigan committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

* commit '556af3f47a96b32898ab4cdbd65b16486a4871e8':
  Add Gradle support

ruiyigan committed:
Use Gradle to automate some of the build tasks of the project like Checkstyle

ruiyigan committed:
Fix java styles based on Checkstyle

ruiyigan committed:
Add JUnit tests to test the behavior of the code.

ruiyigan committed:
Switch to save on any change instead of at the end

ruiyigan committed:
Add for Level 8 branch

ruiyigan committed:
Merge branch 'branch-Level-8'

ruiyigan committed:
Add JavaDoc comments

ruiyigan committed:
Tweak the code to comply with a coding standard

ruiyigan committed:
Give users a way to find a task by searching for a keyword

ruiyigan committed:
Merge branch 'branch-Level-9'

ruiyigan committed:
Merge branch 'branch-A-CodingStandard'

ruiyigan committed:
Merge tag 'branch-A-JavaDoc'

ruiyigan committed:
Create Launcher class

ruiyigan committed:
Add GUI for Duke

ruiyigan committed:
Abstract response to user input functions to ui class

ruiyigan committed:
Fix minor bug regarding import

Flash161203 commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
```
I feel it would be more readable if the import statements are logically separated (those importing from different packages) by blank lines - in other files as well :)

Flash161203 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,83 @@
+     * @param userInput
+     * @param tasks
+     * @throws DukeException
+     */
```
I feel that a blank line between the method description and parameter section would be better :)
I noticed this in several other places as well...

Flash161203 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+    public LocalDate getDeadlineDate(String userInput) throws DukeException {
+        LocalDate deadlineDate;
+        try {
+            deadlineDate = LocalDate.parse(userInput.split(" ", 2)[1].split(" /by ", 2)[1]);
```
Perhaps break the code into different lines to make it more understandable?

Flash161203 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+            throw new DukeException("The start date and end date of a event cannot be empty. \" +\n"
+                    + "\"Make sure dates are in this format yyyy-mm-dd.");
+        }
+        return new LocalDate[]{eventStartDate, eventEndDate};
```
I think that a whitespace between [] and {} would be nice...

Flash161203 reviewed (commented):
Overall, I found the code to be very neat and compliant with the coding standard 👍 . Just a few minor things to 'fix'. 😄

Flash161203 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+    /**
+     * Different types of command by user
+     */
+    public enum Action {
+        LIST,
+        TODO,
+        DEADLINE,
+        EVENT,
+        MARK,
+        UNMARK,
+        DELETE,
+        FIND,
+        UNKNOWN
+    };
```
I like the fact that you used enums to represent the different commands 👍

ruiyigan replied:
thank you for your feedback!

Cyrus-Krispin commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,97 @@
+     */
+
+    public ArrayList<Task> loadData() throws DukeException {
+        ArrayList<Task> taskList = new ArrayList<>();
+
+        File file = new File(this.filePath);
+        try {
+            // create file if file is empty
+            if (!file.exists()) {
+                file.getParentFile().mkdir();
+                file.createNewFile();
+            }
+            Scanner fileReader = new Scanner(file);
+            Task task;
+            while (fileReader.hasNextLine()) {
+                String data = fileReader.nextLine();
+                String[] dataArray = data.split("\\| ");
+                String eventType = dataArray[0].trim();
+                String isDone = dataArray[1].trim();
+                if (eventType.equals("T")) {
+                    task = new ToDo(dataArray[2].trim());
+                } else if (eventType.equals("D")) {
+                    LocalDate deadline = LocalDate.parse(dataArray[3].trim());
+                    task = new Deadline(dataArray[2].trim(), deadline);
+                } else {
+                    String[] eventDetails = dataArray[3].split("- ");
+                    LocalDate startDate = LocalDate.parse(eventDetails[0].trim());
+                    LocalDate endDate = LocalDate.parse(eventDetails[1].trim());
+                    task = new Event(dataArray[2].trim(), startDate, endDate);
+                }
+                if (isDone.equals("X")) {
+                    task.markAsDone();
+                }
+                taskList.add(task);
+            }
+        } catch (IOException e) {
+            throw new DukeException(e.getMessage());
+        }
+        return taskList;
+    }
```
Maybe decreasing LOC could increase the readability of this method (more than 30 LOC)

ruiyigan replied:
good idea! I should perhaps use some methods from the Parser class.

Cyrus-Krispin commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,131 @@
+        }
+        return s;
+    }
+    /**
+     * Return task details based on keyword by user
+     * @param tasks
+     * @return task details
+     * @throws DukeException
+     */
```
Adding a space between the description and the parameters will better follow the coding standards. I found this issue in all of your Javadocs

Cyrus-Krispin reviewed (commented):
Overall code is well written just minor changes to be made

ruiyigan committed:
Delete extra test

ruiyigan committed:
Merge branch 'branch-Level-10'

ruiyigan committed:
Add assertion to TaskList class

Added assertions to check if task index that is passed into TaskList methods exists during runtime

ruiyigan committed:
Merge pull request #1 from ruiyigan/branch-A-Assertions

Add assertion to TaskList class

ruiyigan committed:
Add spacing between params and java doc

Improve readibility and code quality.

ruiyigan committed:
Remove excess lines

ruiyigan committed:
Add bye command that close application

Added a bye command to action enum. Currently when user enters bye as their input, the application will close

ruiyigan committed:
Add comment to give original author credit

ruiyigan committed:
Merge pull request #2 from ruiyigan/branch-A-CodeQuality

Branch a code quality

ruiyigan committed:
Publish Build Scans from GitHub Actions

ruiyigan committed:
Add GitHub Actions workflow

ruiyigan committed:
Merge pull request #3 from ruiyigan/branch-A-CI

Branch CI

ruiyigan committed:
Amend GitHub Action File

ruiyigan committed:
Amend GitHub Action File to set up Java

ruiyigan committed:
Enable github actions for non mac users

ruiyigan committed:
Add testing of unit testcases with Gradle

ruiyigan committed:
Add new command to see statistics of tasks

ruiyigan committed:
Handle bugs relating to checkstyle

ruiyigan committed:
Packaged main logic

ruiyigan committed:
Make screen resizable

ruiyigan committed:
Add different background colour for error

ruiyigan committed:
Give it personality

ruiyigan committed:
Bug fix

ruiyigan committed:
Allow resizing when text too long

ruiyigan committed:
Amend minor formatting issue

ruiyigan committed:
Add greetings feature

ruiyigan committed:
Add image to docs

ruiyigan committed:
Switch image location to docs

ruiyigan committed:
Remove extra import

ruiyigan committed:
Change image for docs

ruiyigan committed:
Update README.md

ruiyigan committed:
Update README.md

ruiyigan committed:
Update README.md

ruiyigan committed:
Update build name

ruiyigan committed:
Update README.md

ruiyigan committed:
Update README.md

ruiyigan committed:
Update README.md

ruiyigan committed:
Update README.md

ruiyigan committed:
Added java docs

ruiyigan committed:
Merge branch 'master' of https://github.com/ruiyigan/ip

ruiyigan committed:
Update UI.png
